lib/persist-state.c

```diff
@@ -267,6 +314,12 @@ _alloc_value(PersistState *self, guint32 orig_size, gboolean in_use, guint8 vers
   persist_state_unmap_entry(self, self->current_ofs);
 
   self->current_ofs += size + sizeof(PersistValueHeader);
+
+  if (!_check_watermark(self) && !_grow_store(self, self->current_size + PERSIST_FILE_INITIAL_SIZE))
+    {
+      main_loop_exit();
+    }
+
   return result;
```

Comment:
I like this in general, however a more precise error message should be printed in the log, e.g. it should explicitly say that syslog-ng is going to exit because it cannot grow the persist file.

right now, if I understand correctly it only says: "Can't grow the persist file", but no mentioning that it is going to exit because of this.

Also, have you validated that this way even the last messages make it to their destination? (assuming of course that they are writable).

Also, it's not really nice that we are calling back to mainloop.c from here as that's a circular dependency (mainloop manages persist_state after all, so mainloop -> persiststate not the other way around). it should probably be implemented via a callback where mainloop registers.